### DISCUSSION TITLE
SAMD21 upgrade to CMSIS-DAP v2.x

### DISCUSSION THREAD

**Discussion by sebastianschlupp** [ataradov/free-dap#33] [Q&A]
Hi, 
you mention being willing to update other platforms to CMSIS-DAP v2.x if requested (or needed by you). I am designing a small piece of hardware based on the SAMD21 and would like to know if you have the time and interest of updating that platform.

Thank you,
Sebastian

**Comment by ataradov** [comment]
Yeah, sure. It is very easy to do. I'll do it a bit later, but you can have a go at it yourself. It would be basically D11 with header files and a linker script replaced and things adjusted as needed to build.

**Comment by ataradov** [comment]
I've updated it. It was even easier than I expected.
I have not done a lot of testing, but it seems to work and there are not a lot of changes compared to the D11.

**Reply by sebastianschlupp** [reply]
Hi, thanks a lot. I'll gladly try it out with my hardware as soon as I find a bit of free time. This is more of a hobby project and, as you probably know, those are always last on the list of priorities :)

**Reply by ataradov** [reply]
You can also compare the linker files. The only difference is in the " flash (rx) : ORIGIN = 0x00000800, LENGTH = 14K" line where it offsets the code to accommodate the bootloader.

**Reply by Jackywtw** [reply]
Thanks~ I just found that... I tried modified it to reserve 8K for the bootloader.. Now, it can be flashed into and got the DAP v2 adapter now. ^_^

**Reply by Jackywtw** [reply]
Hi, 

I am a newbie for SWD  adapter with samd21... Could you tell me which pins from samd21 provide the SWD port connections?

**Reply by ataradov** [reply]
https://github.com/ataradov/free-dap/blob/master/platform/samd21/hal_config.h

**Comment by sebastianschlupp** [comment]
Finally managed to get time to look into it.
I have started simple attempting to get the debug probe to be seen by pyocd and to erase / reset / flash a device. 
All of the three commands work without issue on a blank target but once programmed neither work anymore. I dug a bit deeper into it and it turns out that the controller never toggles the reset pin.
I'd like to mention that I have used different pins than in your example implementation but if I use the out / set / clear commands defined in dap_config.h, the reset pin toggles without issues, even when connected to the target.
Below you can see the commands and the corresponding traces:

![image](https://github.com/ataradov/free-dap/assets/42110132/463521b4-5b9b-4b2e-a991-55702174067d)
![image](https://github.com/ataradov/free-dap/assets/42110132/cf661929-9821-4db8-abb7-75e81d51ac03)

There seems to be no toggling of the reset pin whatsoever. Of course programming a blank target works even without the reset line but ones the target is running, it can't be reprogrammed/erased without a working reset line.
I will start going through the code and will try to figure out what is happening but I assume that you might be able to point me into the right direction and maybe save some time. Thanks a lot.

**Comment by ataradov** [comment]
The reset is not mandatory unless your firmware changes the functionality of the SWD pins or goes to sleep.

I'm not familiar with pyocd, I don't know if it does anything with reset. Debugger itself does not use it, it is up to the host to control it.

Try with my edbg tool (https://github.com/ataradov/edbg) and see if it works that way
